### PR TITLE
[Sema] Synthesize pre-type-checked ASTs for more derived conformances

### PR DESCRIPTION
**Pull request by DougGregor**
Synthesize pre-type-checked ASTs for a few places where we are deriving conformances:
* The getter for `hashValue`
* `==` for uninhabited enums
* `==` for enums whose cases have no associated values

Should fix rdar://problem/54712316, a case where the non-type-checked
AST was getting processed by SILGen.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/Sema/DerivedConformanceEquatableHashable.cpp [hunk 1]
```diff
@@ -1132,19 +1160,49 @@ deriveBodyHashable_hashValue(AbstractFunctionDecl *hashValueDecl, void *) {
...
+                                     /*implicit*/ true,
+                                     AccessSemantics::Ordinary,
+                                     selfType);
+
+  // _hashValue(for:)
+  auto *hashFunc = C.getHashValueForDecl();
+  if (!hashFunc->hasInterfaceType())
+    static_cast<TypeChecker *>(C.getLazyResolver())->validateDecl(hashFunc);
+
+  auto substitutions = SubstitutionMap::get(
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: This should get SubstitutionMap::getProtocolSubstitutions()

Comment (DougGregor) [reply lib/Sema/DerivedConformanceEquatableHashable.cpp]: Ah, thank you!

Comment (DougGregor) [reply lib/Sema/DerivedConformanceEquatableHashable.cpp]: No, never mind: it's not a protocol method, it's a free function.

**Pull request merged**